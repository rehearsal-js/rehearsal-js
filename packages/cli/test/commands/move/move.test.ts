import { resolve } from 'node:path';
import { afterEach, beforeEach, describe, expect, test } from 'vitest';
import { Project } from 'fixturify-project';
import fastGlob from 'fast-glob';
import { cleanOutput, prepareProject, runBin } from '../../test-helpers/index.js';

function sanitizeAbsPath(basePath: string, files: string[]): string[] {
  return files.map((file) => file.replace(basePath, ''));
}

describe('Command: move', () => {
  let project: Project;

  beforeEach(async () => {
    project = prepareProject('base_js_app');
    await project.write();
  });

  afterEach(() => {
    project.dispose();
  });

  test('move file', async () => {
    const sourceDir = 'src/foo';

    const result = await runBin('move', [`${sourceDir}/baz.js`], ['--rootPath', project.baseDir], {
      cwd: project.baseDir,
    });
    const projectSourceDir = resolve(project.baseDir, sourceDir);

    const jsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{js,gjs}`, {
      cwd: project.baseDir,
    });
    const tsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{ts,gts}`, {
      cwd: project.baseDir,
    });

    expect(cleanOutput(result.stdout, project.baseDir)).toMatchSnapshot();
    expect(jsSourceFiles).length(3);
    expect(tsSourceFiles).length(1);
    expect(sanitizeAbsPath(projectSourceDir, tsSourceFiles)).toMatchObject(['/baz.ts']);
  });

  test('move dir and sub-dir', async () => {
    const sourceDir = 'src';
    const projectSourceDir = resolve(project.baseDir, sourceDir);

    const tsSourceFilesBeforeMove = fastGlob.sync(`${projectSourceDir}/**/*.{ts,gts}`, {
      cwd: project.baseDir,
    });

    expect(tsSourceFilesBeforeMove).length(1);

    const result = await runBin('move', [`${sourceDir}`], ['--rootPath', project.baseDir], {
      cwd: project.baseDir,
    });

    // check for js and gjs files -> ts and gts files
    const jsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{js,gjs}`, {
      cwd: project.baseDir,
    });
    const tsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{ts,gts}`, {
      cwd: project.baseDir,
    });

    expect(cleanOutput(result.stdout, project.baseDir)).toMatchSnapshot();
    expect(jsSourceFiles).length(0);
    expect(tsSourceFiles).length(6);
    expect(sanitizeAbsPath(project.baseDir, tsSourceFiles)).toMatchObject([
      '/src/bizz.ts',
      '/src/index.ts',
      '/src/foo/baz.ts',
      '/src/foo/biz.ts',
      '/src/foo/e.gts',
      '/src/foo/buz/biz.ts',
    ]);
  });

  test('move package with --graph and --deps flag', async () => {
    const childPackage = 'module-a';

    const result = await runBin(
      'move',
      [`${childPackage}`],
      ['--graph', '--deps', '--rootPath', project.baseDir],
      {
        cwd: project.baseDir,
      }
    );

    const projectSourceDir = resolve(project.baseDir, childPackage);
    // check for js and gjs files -> ts and gts files
    const jsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{js,gjs}`, {
      cwd: project.baseDir,
    });
    const tsSourceFiles = fastGlob.sync(`${projectSourceDir}/**/*.{ts,gts}`, {
      cwd: project.baseDir,
    });

    expect(cleanOutput(result.stdout, project.baseDir)).toMatchSnapshot();
    expect(jsSourceFiles).length(0);
    expect(tsSourceFiles).length(3);
    expect(sanitizeAbsPath(project.baseDir, tsSourceFiles)).toMatchObject([
      '/module-a/index.ts',
      '/module-a/src/baz.ts',
      '/module-a/src/foo.ts',
    ]);
  });

  test('expect failure when --deps is passed without --graph', async () => {
    const childPackage = 'module-a';

    await expect(
      async () =>
        await runBin('move', [`${childPackage}`], ['--deps', '--rootPath', project.baseDir], {
          cwd: project.baseDir,
        })
    ).rejects.toThrowError(`'--deps' can only be passed when you pass --graph`);
  });

  test('expect failure when --devDeps is passed without --graph', async () => {
    const childPackage = 'module-a';
    await expect(
      async () =>
        await runBin('move', [`${childPackage}`], ['--devDeps', '--rootPath', project.baseDir], {
          cwd: project.baseDir,
        })
    ).rejects.toThrowError(`'--devDeps' can only be passed when you pass --graph`);
  });

  test('expect failure when --ignore is passed without --graph', async () => {
    const childPackage = 'module-a';
    await expect(
      async () =>
        await runBin(
          'move',
          [`${childPackage}`],
          ['--ignore', 'foo', '--rootPath', project.baseDir],
          {
            cwd: project.baseDir,
          }
        )
    ).rejects.toThrowError(`'--ignore' can only be passed when you pass --graph`);
  });
});
