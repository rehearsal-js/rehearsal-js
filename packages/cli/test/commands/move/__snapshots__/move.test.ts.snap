// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Command: move > can opt-out of move sub directory with graph 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[DATA] ./src/foo/biz.js -> ./src/foo/biz.ts
[DATA] ./src/foo/e.gjs -> ./src/foo/e.gts
[DATA] ./src/foo/buz/biz.js -> ./src/foo/buz/biz.ts
[SUCCESS] Executing git mv"
`;

exports[`Command: move > can opt-out of move with graph 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/index.js -> ./src/index.ts
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[DATA] ./src/foo/biz.js -> ./src/foo/biz.ts
[DATA] ./src/foo/e.gjs -> ./src/foo/e.gts
[DATA] ./src/foo/buz/biz.js -> ./src/foo/buz/biz.ts
[SUCCESS] Executing git mv"
`;

exports[`Command: move > move dir and sub-dir 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/index.js -> ./src/index.ts
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[DATA] ./src/foo/biz.js -> ./src/foo/biz.ts
[DATA] ./src/foo/e.gjs -> ./src/foo/e.gts
[DATA] ./src/foo/buz/biz.js -> ./src/foo/buz/biz.ts
[SUCCESS] Executing git mv"
`;

exports[`Command: move > move file 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[SUCCESS] Executing git mv"
`;

exports[`Command: move > move packages based on the graph 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/index.js -> ./src/index.ts
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[DATA] ./src/foo/biz.js -> ./src/foo/biz.ts
[DATA] ./src/foo/e.gjs -> ./src/foo/e.gts
[DATA] ./src/foo/buz/biz.js -> ./src/foo/buz/biz.ts
[DATA] ./module-a/src/baz.js -> ./module-a/src/baz.ts
[DATA] ./module-a/src/foo.js -> ./module-a/src/foo.ts
[DATA] ./module-a/src/index.js -> ./module-a/src/index.ts
[DATA] ./module-b/src/tires.js -> ./module-b/src/tires.ts
[DATA] ./module-b/src/car.js -> ./module-b/src/car.ts
[DATA] ./module-b/src/index.js -> ./module-b/src/index.ts
[SUCCESS] Executing git mv"
`;

exports[`Command: move > move sub directory based on the graph 1`] = `
"info:    @rehearsal/move<test-version>
[STARTED] Analyzing project dependency graph
[SUCCESS] Analyzing project dependency graph
[STARTED] Executing git mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] git mv failed, using mv
[DATA] renamed: 
[DATA] ./src/foo/baz.js -> ./src/foo/baz.ts
[DATA] ./src/foo/biz.js -> ./src/foo/biz.ts
[DATA] ./src/foo/e.gjs -> ./src/foo/e.gts
[DATA] ./src/foo/buz/biz.js -> ./src/foo/buz/biz.ts
[SUCCESS] Executing git mv"
`;
